### PR TITLE
IE9 fixes - to make SVG maps compatible

### PR DESCRIPTION
**Pull request by alexcjohnson**
A bunch of random fixes for IE9 behavior, only one was actually *in* maps, oddly enough (so some edge cases may have been broken in other plot types too), but all of which are *accessed by* maps.

- ~~`classList`~~
- ~~`instanceof d3.selection`~~ -> seems to work for built-in types and some user-defined types, but for some reason `d3.selection` broke it, replaced with duck typing
- ~~`console[fn].apply`~~ - I guess these aren't regular functions in IE9, so if there's no `apply` just call the function once for each argument.
- When an invalid attribute value is provided, IE9 throws an error rather than just using a default value. Choropleth with no explicit line width was trying to set `undefinedpx` via `drawing.dashLine`, which in all other browsers silently defaulted to `1px`. I made this explicit at the `supplyDefaults` level, while at the same time defaulting to `0px` in `dashLine` so this can't cause further problems.

cc @chriddyp @etpinard 

Given the scattered nature of these fixes, it's probably worthwhile to have someone spend some time manually trying out more plots & behaviors, ideally on a real local copy of IE9 rather than Browserstack. I saw some sporadic glitches with drag interactions, including one where it seemed like the hover effects were pointing to the old point locations... but couldn't reproduce so I'm concerned it was just an artifact of tunneling through Browserstack.

### CODE REVIEW

File: tasks/test_syntax.js [hunk 1]
```diff
@@ -56,7 +60,15 @@ function assertHeaders() {
 
             // parse through code string while keeping track of comments
             var comments = [];
-            falafel(code, {onComment: comments, locations: true}, function() {});
+            falafel(code, {onComment: comments, locations: true}, function(node) {
+                // look for .classList
+                if(node.type === 'MemberExpression') {
+                    var parts = node.source().split('.');
+                    if(parts[parts.length - 1] === 'classList') {
+                        logs.push(file + ' : contains .classList (IE failure)');
```
Comment (alexcjohnson) [inline tasks/test_syntax.js]: I was hoping eslint would have an option for IE compatibility (actually not even IE11 supports `classList` on SVG elements, apparently)... but I don't see one, so I added it in here. Confirmed that before the change I made to `modebar.js` it failed this test.

File: test/jasmine/tests/lib_test.js [hunk 2]
```diff
@@ -1599,6 +1600,172 @@ describe('Test lib.js:', function() {
...
+    });
+
+    describe('loggers', function() {
+        var stashConsole,
+            stashLogLevel;
+
+        function consoleFn(name, hasApply, messages) {
+            var out = function() {
+                var args = [];
+                for(var i = 0; i < arguments.length; i++) args.push(arguments[i]);
```
Comment (alexcjohnson) [inline test/jasmine/tests/lib_test.js]: because `toEqual` doesn't think `arguments` is a regular array

File: tasks/test_syntax.js (hunk 1 above)
Comment (chriddyp) [inline tasks/test_syntax.js]: Great idea!

File: src/components/drawing/index.js [hunk 3]
```diff
@@ -112,6 +112,7 @@ drawing.lineGroupStyle = function(s, lw, lc, ld) {
 };
 
 drawing.dashLine = function(s, dash, lineWidth) {
+    lineWidth = +lineWidth || 0;
```
Comment (etpinard) [inline src/components/drawing/index.js]: Interesting. Maybe this also fixes https://github.com/plotly/plotly.js/issues/166 ?

Comment (alexcjohnson) [inline src/components/drawing/index.js]: Maybe - I'll try that tonight and see if it implies any further fixes. Any other IE issues I should include in my testing?

Comment (etpinard) [inline src/components/drawing/index.js]: Maybe

- https://github.com/plotly/plotly.js/issues/1247
- https://github.com/plotly/plotly.js/issues/746

Comment (alexcjohnson) [inline src/components/drawing/index.js]: I think this does fix #166 - and https://github.com/plotly/plotly.js/pull/1332/commits/4aa1a4535a63633f38256159fc751ca59dd7b2bd should catch any of these `nanpx` or `undefinedpx` errors, at least as long as they go through d3.

Will investigate the other two tomorrow, they may be related.

File: test/image/index.html [hunk 4]
```diff
@@ -5,6 +5,7 @@
 <script type="text/javascript" src="../plotly.js/dist/extras/mathjax/MathJax.js?config=TeX-AMS-MML_SVG"></script>
 <script type="text/javascript" src="../plotly.js/build/plotly.js" charset="utf-8"></script>
 <script type="text/javascript" src="../plotly.js/dist/plotly-geo-assets.js" charset="utf-8"></script>
+<script type="text/javascript" src="../plotly.js/test/image/strict-d3.js" charset="utf-8"></script>
```
Comment (etpinard) [inline test/image/index.html]: Great idea here!

First, I was thinking adding this to the image test index might lead to some very annoying debugging sessions when things go wrong. But, by adding it to the test-dashboard as well essentially make my concerns disappear  (except for devs that don't use the test-dashboard cc @rreusser @monfera).

Alternatively, we could mock the `d3.selection.prototype` only in the [`ie9_test.js`](https://github.com/plotly/plotly.js/blob/master/test/jasmine/bundle_tests/ie9_test.js) bundle test and test each ie9 falback one by one.

It comes down to choosing between a solution with great coverage (yours) and a less intrusive solution that might lead to less :face_with_head_bandage: down the road. I'm ok with both.

Comment (alexcjohnson) [inline test/image/index.html]: We've talked before about wanting a way to report errors out of the image tester - printing the stack trace to a blank image or something. My vote would be to leave this in and work toward that goal. Being strict here will help us even outside IE, as we really never want to implicitly rely on browser defaults.

File: src/lib/index.js [hunk 5]
```diff
@@ -451,8 +451,12 @@ lib.addStyleRule = function(selector, styleString) {
     else lib.warn('addStyleRule failed');
 };
 
-lib.getTranslate = function(element) {
+// TODO (alexcjohnson): the following translate/scale functions should
+// get moved to components/drawing rather than the generic lib.
```
Comment (alexcjohnson) [inline src/lib/index.js]: I didn't want to move all of these (along with all their references and tests) without a discussion, but I did write a new one before finding this in here and discovering that it already tests the behavior I needed it to have for IE.

File: src/lib/svg_text_utils.js [hunk 6]
```diff
@@ -278,6 +280,13 @@ function encodeForHTML(_str) {
 function convertToSVG(_str) {
     _str = convertEntities(_str);
 
+    // normalize behavior between IE and others wrt newlines and whitespace:pre
+    // this combination makes IE barf https://github.com/plotly/plotly.js/issues/746
+    // Chrome and FF display \n, \r, or \r\n as a space in this mode.
+    // I feel like at some point we turned these into <br> but currently we don't so
+    // I'm just going to cement what we do now in Chrome and FF
+    _str = _str.replace(NEWLINES, ' ');
```
Comment (alexcjohnson) [inline src/lib/svg_text_utils.js]: As far as I can tell this is purely a bug in IE - though one that persists from 9 through 11... the affected text is invisible, even though `getBoundingClientRect()` on its node shows the correct position and shape, and it reappears when you take out the newline character.

File: src/lib/index.js (hunk 5 above)
Comment (etpinard) [inline src/lib/index.js]: :+1: for moving `getTranslate` and `setTranslate` out of `Lib`. Maybe putting them into `Drawing` would make the most sense?

File: src/lib/svg_text_utils.js (hunk 6 above)
Comment (etpinard) [inline src/lib/svg_text_utils.js]: :1st_place_medal: debugging.

File: test/jasmine/tests/hover_label_test.js [hunk 7]
```diff
@@ -123,7 +125,8 @@ describe('hover info', function() {
 
             expect(d3.selectAll('g.axistext').size()).toEqual(0);
             expect(d3.selectAll('g.hovertext').size()).toEqual(1);
-            expect(d3.selectAll('g.hovertext').select('text').html()).toEqual('hover text');
+            expect(d3.selectAll('g.hovertext').select('text').html())
+                .toEqual('hover text  with spaces  not newlines');
```
Comment (alexcjohnson) [inline test/jasmine/tests/hover_label_test.js]: None of the image tests in https://github.com/plotly/plotly.js/pull/1332/commits/21059303ee919812e4b2e281df5f71f7c77cf7be actually changed results with the fix there, but they would all have resulted in invisible text in IE before the fix. But this test confirms that we do sanitize newlines for IE compatibility.

Comment (etpinard) [inline test/jasmine/tests/hover_label_test.js]: Nicely done.

File: src/lib/index.js (hunk 5 above)
Comment (alexcjohnson) [inline src/lib/index.js]: Yeah, that's where I put mine before noticing these. Great, I'll move them in there.

Comment (alexcjohnson) [inline src/lib/index.js]: done in https://github.com/plotly/plotly.js/pull/1332/commits/66baf516524d114ba2044b2f432a3c3441c5ea43

Any final comments @etpinard ?

**Comment by etpinard** [comment]
:dancer:

**Pull request merged**